tactical-microgrid-standard/common/Controller.cpp

```diff
@@ -0,0 +1,59 @@
+#include "Controller.h"
+
+DDS::ReturnCode_t Controller::init(DDS::DomainId_t domain, int argc, char* argv[])
+{
+  DDS::ReturnCode_t rc = join_domain(domain, argc, argv);
+  if (rc != DDS::RETCODE_OK) {
+    return rc;
+  }
+
+  rc = create_subscribers(
+    [&](const tms::DeviceInfo& di, const DDS::SampleInfo& si) {
+      got_device_info(di, si);
+    },
+    [&](const tms::Heartbeat& hb, const DDS::SampleInfo& si) {
+      got_heartbeat(hb, si);
+    }
+  );
```

Comment:
Nevermind, I think directly passing non-static member function doesn't work.

Reply:
> Directly pass `got_device_info` and `got_heartbeat` without the lambda wrappers?

Didn't occur to me...

> https://en.cppreference.com/w/cpp/utility/functional/mem_fn may help

Sounds like a bound method in Python which is what you'd want in this situation.

Reply:
`std::mem_fn` won't work because it requires the object as the first argument when calling it, so it's not a bound method like I thought at first.

[`std::bind`](https://en.cppreference.com/w/cpp/utility/functional/bind) is almost the same thing as a Python bound method, but it binds both the object and the arguments so you have to pass these placeholders if you want them to be parameters on the callable result:

```c++
  using namespace std::placeholders;
  rc = create_subscribers(
    std::bind(&Controller::got_device_info, this, _1, _2),
    std::bind(&Controller::got_heartbeat, this, _1, _2));
```

It's pre-C++11, so that's neat I guess because I'm not sure I was aware of it before.

C++20 has [`std::bind_front`](https://en.cppreference.com/w/cpp/utility/functional/bind_front), which doesn't require the placeholders.

Reply:
Right, `mem_fun` doesn't do partial application.  `bind` can do it with C++11 (or boost/tr1) but it's verbose.  I tend to go with lambdas for these cases, I think it's easier to read.

Reply:
You may want to try lambdas with `auto` parameters (C++14).